### ISSUE TITLE
Incompatible with Embroider 's `staticAddonTestSupportTrees` optimized mode

### ISSUE THREAD

**Issue by patocallaghan** [san650/ember-cli-page-object#539]
I've been testing out various addons we use in our app against Embroider and I've found an incompatibility between it and ember-cli-page-object.

### Reproduction
Just setting up a simple page object and trying to use it in a test causes for the following test error:

> Looks like you attempt to access page object property outside of test context. If that's not the case, please make sure you use the latest version of "@ember/test-helpers"

![image](https://user-images.githubusercontent.com/685034/107544816-84286780-6bc2-11eb-8ee9-f1b037118bc1.png)

```js
// tests/pages/users.js
import {
  create,
  visitable
} from 'ember-cli-page-object';

export default create({
  visit: visitable('/')
});
```

```js
// tests/acceptance/index-test.js
import { module, test } from 'qunit';
import { visit, currentURL } from '@ember/test-helpers';
import { setupApplicationTest } from 'ember-qunit';
import page from 'embroider-spike/tests/pages/users';

module('Acceptance | index', function (hooks) {
  setupApplicationTest(hooks);

  test('visiting /index', async function (assert) {
    await page.visit();

    assert.equal(currentURL(), '/');
  });
});
```

I have a branch [here](https://github.com/GavinJoyce/embroider-spike/compare/master...patocallaghan:patoc/ember-cli-page-object) which recreates the issue.

* Run `yarn test:ember -s -f=visiting`. This uses Embroider and the test fails ❌
* Run `CLASSIC=true yarn test:ember -s -f=visiting`. This uses the existing build pipeline and test passes ✅

### Problem
The problem looks to stem from when you enable Embroider's `staticAddonTestSupportTrees` setting. 

From Embroider's [README](https://github.com/embroider-build/embroider#options)

> Enable `staticAddonTestSupportTrees` and `staticAddonTrees` and test your application. This is usually safe, because most code in these trees gets consumed via import statements that we can analyze. But you might find exceptional cases where some code is doing a more dynamic thing.

If I set this setting to `false` the tests pass fine.

The error mentions to make sure to have the latest version of `@ember/test-helpers` but doing a check it appears I am:

```
> yarn list ember-qunit
└─ ember-qunit@5.1.2
```

```
> yarn list @ember/test-helpers
└─ @ember/test-helpers@2.1.4
```

### Some other info
I tried setting the context by calling `page.setContext(this)` in the test. This progresses the test but it then fails for another reason.

> Ember.emberDeprecate is not a function

![image](https://user-images.githubusercontent.com/685034/107547223-1e89aa80-6bc5-11eb-90ba-a8cb91a363ae.png)

I don't think this is related to ember-cli-page-object though. I think there's an issue with the aliasing of the function to `emberDeprecate`. I've been able to replicate this exact thing in app code outside the context of ember-cli-page-object. I'm going to log an issue for that elsewhere.

**Comment by alexlafroscia** [comment]
That later issue (the one about `emberDeprecate` is something I ran into as well -- #540 should fix that part!

**Comment by ro0gr** [comment]
Well, the day has come 🙂  Thanks for opening the Embroider topic!

I meant to test against it for a quite a long time, but didn't really know how to do it. I suspect an old ember-cli-babel@6 would be one of the first blocker for that. I'd give it a try by adding an ember-try scenario(ispired by https://github.com/ember-cli/ember-cli/pull/9436), let's see how far we can go in scope of ec-page-object@v1.

**Comment by patocallaghan** [comment]
@ro0gr no problem. I tried to initially put together a PR which added Embroider ember-try scenarios but it failed spectacularly 😃 As you mention it looked to be because of the old ember-cli/ember-cli-babel etc and I didn't want to boil the ocean 😅

@alexlafroscia thanks for figuring out the `emberDeprecate` in https://github.com/san650/ember-cli-page-object/pull/540 🙌

**Comment by ro0gr** [comment]
A quick update. I've started looking into this. The first blocker appeared is TS tests. Created a ticket for this(https://github.com/embroider-build/embroider/issues/691), Hopefuly, it'd be easy to fix.

**Comment by ro0gr** [comment]
Keep pushing it slowly. I think I'd like to make it work on the v2 branch first, cause there should be just a bit less magic, and if there are some blockers, we're a bit safer to adjust it there.

Currently, it's blocked with some issues related to`buildEmberPlugins(`, which we've started to use in the #531. Reported it to the ember-cli-babel for now(https://github.com/babel/ember-cli-babel/issues/398).

**Comment by rwjblue** [comment]
I just debugged this with @thoover and @krisselden (working on one of our internal applications). The basic issue is that this code:

https://github.com/san650/ember-cli-page-object/blob/82b7e85c9c45d0cf5cf6fe5f95f2287dbd1a7685/addon-test-support/-private/compatibility.js#L43-L51

Is not understood by Embroider (with that staticAddonTestSupportTrees flag set to `true`). We end up hitting the else case in this conditional (and any others that use `require.has` like this) and then not finding the associated context.

**Comment by rwjblue** [comment]
The thing I'm not sure about here are if we actually have the same support constraints as were in place when this compatibility layer was introduced?

**Comment by rwjblue** [comment]
I think the things that are being done there could be implemented in terms of `@embroider/macros` 🤔

**Comment by ro0gr** [comment]
@rwjblue sorry for inconveniences. It takes a some longer to move it to the finish line, but we are going to solve it soon.

> I think the things that are being done there could be implemented in terms of @embroider/macros 🤔

Maybe it could, though I haven't tried it and not sure if it requires `ember-auto-import@2` to be installed to make it work(which would be a breaking change). In case if it's possible to do in a backward compatible way I'd be glad to accept such a PR. However, TBH, I suspect there might appear some other roadblocks for the proper embroider support, related to some tricks in the https://github.com/san650/ember-cli-page-object/blob/master/index.js, or what else stuff. 

The current plan is to remove this compatibility layer completely for the next major release of ec-page-object, and make embroider supported properly.

**Comment by ro0gr** [comment]
This should be addressed now in the [v2.0.0-beta.3](https://github.com/san650/ember-cli-page-object/releases/tag/v2.0.0-beta.3)
